In [1]:
# standard imports
import pandas as pd
import numpy as np

# Acquire data sets

In [2]:
# read in 2010 San Antonio census data
df_10 = pd.read_csv('ACS-1YR-2010.csv')
# read in 2011 San Antonio census data
df_11 = pd.read_csv('ACS-1YR-2011.csv')
# read in 2012 San Antonio census data
df_12 = pd.read_csv('ACS-1YR-2012.csv')
# read in 2013 San Antonio census data
df_13 = pd.read_csv('ACS-1YR-2013.csv')
# read in 2014 San Antonio census data
df_14 = pd.read_csv('ACS-1YR-2014.csv')
# read in 2015 San Antonio census data
df_15 = pd.read_csv('ACS-1YR-2015.csv')
# read in 2016 San Antonio census data
df_16 = pd.read_csv('ACS-1YR-2016.csv')
# read in 2017 San Antonio census data
df_17 = pd.read_csv('ACS-1YR-2017.csv')
# read in 2018 San Antonio census data
df_18 = pd.read_csv('ACS-1YR-2018.csv')
# read in 2019 San Antonio census data
df_19 = pd.read_csv('ACS-1YR-2019.csv')
# read in 2020 San Antonio census data
df_20 = pd.read_csv('ACS-5YR-2020.csv')
# read in 2021 San Antonio census data
df_21 = pd.read_csv('ACS-1YR-2021.csv')

# Prepare

To neatly combine data frames I want to see how many columns and rows each data frame contains. If they all contain the same number of columns and rows, I may be able to concat them together and then clean the single combined data set instead of multiple smaller data sets.

In [3]:
# find the number of columns in each data frame
# list of dataframes
dataframes = [df_10, df_11, df_12, df_13, df_14, 
              df_15, df_16, df_17, df_18, df_19, 
              df_20, df_21]
# for loop
for df in dataframes:
    # print the number of columns in each DataFrame
    num_columns = len(df.columns)
    print(f"Number of columns in the DataFrame: {num_columns}")

Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9
Number of columns in the DataFrame: 9


In [4]:
# find the number of rows in each data frame
# list of dataframes
dataframes = [df_10, df_11, df_12, df_13, df_14, 
              df_15, df_16, df_17, df_18, df_19, 
              df_20, df_21]
# for loop
for df in dataframes:
    # print the number of rows in each DataFrame
    num_rows = len(df)
    print(f"Number of rows in the DataFrame: {num_rows}")

Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145
Number of rows in the DataFrame: 145


My data sets contain the same number of columns and rows. I will proceed with attempting to combine data frames. Because I anticipate the columns to be the same, I will concat the dataframes.

In [9]:
# concat df_10 and df_11
combined_df = pd.concat([df_10, df_11])
combined_df.shape

(290, 11)

When I attempt to concat df_10 and df_11 the data frames combine to 290 rows and 11 columns instead of the expected 145 rows and 9 columns. From a quick but thorough look at the csv files, all the columns and rows should match. It is possible they are not being read in the same? I will clean each data set individualy then combine into a single dataframe.

In [17]:
df_10.head()

Texas!!Estimate  \
Label (Grouping)                                   
EMPLOYMENT STATUS                            NaN   
    Population 16 years and over      19,123,208   
    In labor force                    12,465,332   
        Civilian labor force          12,363,612   
            Employed                  11,271,851   

                                 Texas!!Estimate Margin of Error  \
Label (Grouping)                                                   
EMPLOYMENT STATUS                                            NaN   
    Population 16 years and over                         ±10,217   
    In labor force                                       ±33,691   
        Civilian labor force                             ±33,592   
            Employed                                     ±36,125   

                                 Texas!!Percent  \
Label (Grouping)                                  
EMPLOYMENT STATUS                           NaN   
    Population 16 years and over     19,123,208   
    In labor force                       65.20%   
        Civilian labor force             64.70%   
            Employed                     58.90%   

                                 Texas!!Percent Margin of Error  \
Label (Grouping)                                                  
EMPLOYMENT STATUS                                           NaN   
    Population 16 years and over                            (X)   
    In labor force                                         ±0.2   
        Civilian labor force                               ±0.2   
            Employed                                       ±0.2   

                                 San Antonio city, Texas!!Estimate  \
Label (Grouping)                                                     
EMPLOYMENT STATUS                                              NaN   
    Population 16 years and over                         1,019,273   
    In labor force                                         660,424   
        Civilian labor force                               648,636   
            Employed                                       589,877   

                                 San Antonio city, Texas!!Estimate Margin of Error  \
Label (Grouping)                                                                     
EMPLOYMENT STATUS                                                              NaN   
    Population 16 years and over                                            ±5,093   
    In labor force                                                          ±7,817   
        Civilian labor force                                                ±8,332   
            Employed                                                        ±8,640   

                                 San Antonio city, Texas!!Percent  \
Label (Grouping)                                                    
EMPLOYMENT STATUS                                             NaN   
    Population 16 years and over                        1,019,273   
    In labor force                                         64.80%   
        Civilian labor force                               63.60%   
            Employed                                       57.90%   

                                 San Antonio city, Texas!!Percent Margin of Error  
Label (Grouping)                                                                   
EMPLOYMENT STATUS                                                             NaN  
    Population 16 years and over                                              (X)  
    In labor force                                                           ±0.7  
        Civilian labor force                                                 ±0.7  
            Employed                                                         ±0.8

The data sets have a x-axis label but the label was read in as a column. The column: Label (Grouping), has been set as the index.

In [3]:
# set the index of each data frame to be column: 'Label (Grouping)'
df_10 = df_10.set_index('Label (Grouping)')
df_11 = df_11.set_index('Label (Grouping)')
df_12 = df_12.set_index('Label (Grouping)')
df_13 = df_13.set_index('Label (Grouping)')
df_14 = df_14.set_index('Label (Grouping)')
df_15 = df_15.set_index('Label (Grouping)')
df_16 = df_16.set_index('Label (Grouping)')
df_17 = df_17.set_index('Label (Grouping)')
df_18 = df_18.set_index('Label (Grouping)')
df_19 = df_19.set_index('Label (Grouping)')
df_20 = df_20.set_index('Label (Grouping)')
df_21 = df_21.set_index('Label (Grouping)')

Several changes need to be made to column names for readablity and simplicity. I will be formating column names to be lowercase, snakecase, abreviated, and I will remove any unecessary charachters.

In [18]:
# format column names to be lowercase
df_10.columns = df_10.columns.str.lower()
# format column names to be snakecase
df_10.columns = df_10.columns.str.replace(' ', '_')
# replace !! with snakecase
df_10.columns = df_10.columns.str.replace('!!', '_')
# remove commas
df_10.columns = df_10.columns.str.replace(',', '')
# shorten texas to tx
df_10.columns = df_10.columns.str.replace('texas', 'tx')
# shorten san antonio to sa
df_10.columns = df_10.columns.str.replace('san_antonio', 'sa')
# shorten margin of error to moe
df_10.columns = df_10.columns.str.replace('margin_of_error', 'moe')
# shorten estimate to est
df_10.columns = df_10.columns.str.replace('estimate', 'est')

There are several labels used to label subgroups in the dataframe but do not contain information its self. I will drop these below.

In [26]:
# drop EMPLOYMENT STATUS
df_10.drop(labels='EMPLOYMENT STATUS', inplace=True)
# drop COMMUTING TO WORK
df_10.drop(labels='COMMUTING TO WORK', inplace=True)
# drop OCCUPATION
df_10.drop(labels='OCCUPATION', inplace=True)
# drop INDUSTRY
df_10.drop(labels='INDUSTRY', inplace=True)
# drop CLASS OF WORKER
df_10.drop(labels='CLASS OF WORKER', inplace=True)
# drop INCOME AND BENEFITS (IN 2010 INFLATION-ADJUSTED DOLLARS)
df_10.drop(labels='INCOME AND BENEFITS (IN 2010 INFLATION-ADJUSTED DOLLARS)', inplace=True)
# drop HEALTH INSURANCE COVERAGE
df_10.drop(labels='HEALTH INSURANCE COVERAGE', inplace=True)
# drop PERCENTAGE OF FAMILIES AND PEOPLE WHOSE INCOME IN THE PAST 12 MONTHS IS BELOW THE POVERTY LEVEL
df_10.drop(labels='PERCENTAGE OF FAMILIES AND PEOPLE WHOSE INCOME IN THE PAST 12 MONTHS IS BELOW THE POVERTY LEVEL', inplace=True)

In [27]:
df_10

tx_est tx_est_moe  \
Label (Grouping)                                                     
    Population 16 years and over             19,123,208    ±10,217   
    In labor force                           12,465,332    ±33,691   
        Civilian labor force                 12,363,612    ±33,592   
            Employed                         11,271,851    ±36,125   
            Unemployed                        1,091,761    ±21,621   
...                                                 ...        ...   
    18 years and over                               (X)        (X)   
        18 to 64 years                              (X)        (X)   
        65 years and over                           (X)        (X)   
    People in families                              (X)        (X)   
    Unrelated individuals 15 years and over         (X)        (X)   

                                             tx_percent tx_percent_moe  \
Label (Grouping)                                                         
    Population 16 years and over             19,123,208            (X)   
    In labor force                               65.20%           ±0.2   
        Civilian labor force                     64.70%           ±0.2   
            Employed                             58.90%           ±0.2   
            Unemployed                            5.70%           ±0.1   
...                                                 ...            ...   
    18 years and over                            14.90%           ±0.2   
        18 to 64 years                           15.60%           ±0.2   
        65 years and over                        10.70%           ±0.4   
    People in families                           16.30%           ±0.2   
    Unrelated individuals 15 years and over      26.80%           ±0.5   

                                            sa_city_tx_est sa_city_tx_est_moe  \
Label (Grouping)                                                                
    Population 16 years and over                 1,019,273             ±5,093   
    In labor force                                 660,424             ±7,817   
        Civilian labor force                       648,636             ±8,332   
            Employed                               589,877             ±8,640   
            Unemployed                              58,759             ±4,905   
...                                                    ...                ...   
    18 years and over                                  (X)                (X)   
        18 to 64 years                                 (X)                (X)   
        65 years and over                              (X)                (X)   
    People in families                                 (X)                (X)   
    Unrelated individuals 15 years and over            (X)                (X)   

                                            sa_city_tx_percent  \
Label (Grouping)                                                 
    Population 16 years and over                     1,019,273   
    In labor force                                      64.80%   
        Civilian labor force                            63.60%   
            Employed                                    57.90%   
            Unemployed                                   5.80%   
...                                                        ...   
    18 years and over                                   15.90%   
        18 to 64 years                                  16.20%   
        65 years and over                               14.00%   
    People in families                                  17.60%   
    Unrelated individuals 15 years and over             26.40%   

                                            sa_city_tx_percent_moe  
Label (Grouping)                                                    
    Population 16 years and over                               (X)  
    In labor force                                     